### PR TITLE
refactor: add @cloudflare/voprf-ts/facade and pass crypto as arg everywhere

### PR DESCRIPTION
**Pull request by sublimator**
Supercedes #44

### Pass Around Crypto As Arg Everywhere

1. Oprf.Group -> hash(...) doesn't belong here! and WebCrypto doesn't support shake256
2. Oprf.Crypto -> circular dependencies!  
3. CryptoImpl.provider ->  better, but ick, a global?
4. Explicit optional provider passing (crypto?: CryptoProvider) -> tests failing, where??
5. Required (crypto: CryptoProvider)-> fixed, but damn that's verbose and breaks api
6. --> !!CryptoProvider via...Rest syntax -> nice, can configure Rest as [CryptoProvider] or [CryptoProvider?]!! <---
7. But why? -> Can make sure passing around the provider everywhere
8. But why? ->  Oprf.withConfig({provider: CryptoNoble) and old Oprf work in same process
9. So what? -> In any case ... 

Toggle [this setting](https://github.com/cloudflare/voprf-ts/pull/49/files#diff-939fbbde9cf1b0707741b13dfd69e906bf2ca839475bb966da809b0bc4420c4aR9) here and try the `npm run examples` in both modes
Fails to build when it's required - [added to ci](https://github.com/cloudflare/voprf-ts/pull/49/files#diff-1db27d93186e46d3b441ece35801b244db8ee144ff1405ca27a163bfe878957fR18)

### @cloudflare/voprf-ts/facade, backwards compat impl of #46

See [examples](https://github.com/sublimator/voprf-ts/blob/nd-switch-flipping-2023-10-10/examples/facade/index.ts#L3-L6)

### Named imports for examples

Uses tsx to execute examples/bench which supports runtime compilerOptions.paths to provide more readily modifiable examples.  

### Breaking changes

Minimum changes needed for: https://github.com/cloudflare/privacypass-ts/pull/15
Also see: [Port to facade](https://github.com/armfazh/privacypass-ts/compare/implVoprfToken2...sublimator:privacypass-ts:nd-facade-charade-2023-10-15?expand=1)

```typescript
diff --git a/src/priv_verif_token.ts b/src/priv_verif_token.ts
index 08c95c9..8809078 100644
--- a/src/priv_verif_token.ts
+++ b/src/priv_verif_token.ts
@@ -9,7 +9,6 @@ import {
     VOPRFClient,
     VOPRFServer,
     generateKeyPair,
-    type DLEQParams,
     type Group,
     type SuiteID,
     type HashID,
@@ -31,7 +30,6 @@ export interface VOPRFExtraParams {
     Ns: number;
     Nk: number;
     hash: HashID;
-    dleqParams: DLEQParams;
 }
 
 const VOPRF_SUITE = Oprf.Suite.P384_SHA384;
@@ -44,12 +42,6 @@ const VOPRF_EXTRA_PARAMS: VOPRFExtraParams = {
     Ns: VOPRF_GROUP.scalarSize(),
     Nk: Oprf.getOprfSize(VOPRF_SUITE),
     hash: VOPRF_HASH,
-    dleqParams: {
-        gg: VOPRF_GROUP,
-        hashID: VOPRF_HASH,
-        hash: Oprf.Crypto.hash,
-        dst: '',
-    },
 } as const;
 
 // Token Type Entry Update:
@@ -82,6 +74,7 @@ export class TokenRequest2 {
     //   } TokenRequest;
 
     tokenType: number;
+
     constructor(
         public readonly truncatedTokenKeyId: number,
         public readonly blindedMsg: Uint8Array,
@@ -257,7 +250,7 @@ export class Client2 {
             throw new Error('no token request was created yet');
         }
 
-        const proof = DLEQProof.deserialize(VOPRF.dleqParams, tokRes.evaluateProof);
+        const proof = DLEQProof.deserialize(VOPRF.group.id, tokRes.evaluateProof);
         const evaluateMsg = VOPRF.group.desElt(tokRes.evaluateMsg);
         const evaluation = new Evaluation(Oprf.Mode.VOPRF, [evaluateMsg], proof);
         const [authenticator] = await this.finData.vClient.finalize(
```

### TODO

- [ ] Full document Security.md that summarizes the current status of both back-ends (@armfazh)

### CODE REVIEW

File: src/facade/types.ts [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    finalize: M extends MODEMAP['poprf']
+        ? (
+              finData: FinalizeData,
+              evaluation: Evaluation,
+              info?: Uint8Array
+          ) => Promise<Array<Uint8Array>>
+        : (finData: FinalizeData, evaluation: Evaluation) => Promise<Array<Uint8Array>>
+}
+
+export interface Server<M extends ModeID = ModeID, S extends SuiteID = SuiteID>
```
Comment (sublimator) [inline src/facade/types.ts:83]: Q: TODO: SuiteID even really of any use ?

File: package.json [hunk 2]
```diff
@@ -1,81 +1,91 @@
...
+    "build": "tsc -b . tsconfig.cjs.json",
+    "clean": "tsc -b --clean . test ./tsconfig.cjs.json test/tsconfig.cjs.json bench examples",
+    "test": "npm run test:esm && npm run test:cjs",
+    "test:esm": "npm run test:build && node --experimental-vm-modules node_modules/.bin/jest --ci --selectProjects esm --coverageDirectory coverage/esm",
+    "test:cjs": "npm run test:build && node node_modules/.bin/jest --ci --selectProjects cjs --coverageDirectory coverage/cjs",
+    "test:build": "tsc -b test test/tsconfig.cjs.json",
+    "examples": "tsx examples/index.ts",
+    "examples:facade": "tsx examples/facade/index.ts",
+    "lint": "eslint .",
+    "bench": "tsx bench/index.ts",
```
Comment (sublimator) [inline package.json:83]: using npx for examples/bench

File: examples/tsconfig.json [hunk 3]
```diff
@@ -1,8 +1,13 @@
 {
     "extends": "../tsconfig.json",
-    "include": [
-        "."
-    ],
+    "compilerOptions": {
+        "paths": {
+            "@cloudflare/voprf-ts": ["../src/index.js"],
```
Comment (sublimator) [inline examples/tsconfig.json:5]: makes examples easily copy/pastable

File: src/sjcl/index.js [hunk 4]
```diff
@@ -3599,12 +3599,12 @@ sjcl.codec.arrayBuffer = {
...
 
-if(typeof module !== 'undefined' && module.exports){
-  module.exports = sjcl;
-}
-if (typeof define === "function") {
-    define([], function () {
-        return sjcl;
-    });
-}
+// if(typeof module !== 'undefined' && module.exports){
```
Comment (sublimator) [inline src/sjcl/index.js:3602]: this seems unneeded - was done to stop tsx complaining

File: src/groupTypes.ts [hunk 5]
```diff
@@ -98,6 +87,8 @@ export interface Group extends SerializationHelpers {
 }
 
 export interface GroupCons {
-    fromID(id: GroupID): Group
+    get(id: GroupID): Group
```
Comment (sublimator) [inline src/groupTypes.ts:90]: idempotent/get/cached semantics

File: src/facade/impl.ts [hunk 6]
```diff
@@ -0,0 +1,124 @@
+/* eslint-disable @typescript-eslint/no-unused-vars,@typescript-eslint/no-explicit-any */
+// noinspection JSUnusedLocalSymbols
+import * as assert from 'assert'
```
Comment (armfazh) [inline src/facade/impl.ts:3]: probably not needed, a simply `if` can make it

**Comment by sublimator** [comment]
On mobile - I guess using assert means you need a bundler / polyfill

File: src/facade/types.ts [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+export interface UsesCrypto {
+    readonly crypto: CryptoProvider
+    readonly group: Group
+}
+
+export type DLEQProof = { c: Scalar; s: Scalar; serialize(): Uint8Array }
+// OPRF protocol only specifies encodings for these elements
+export type OPRFElement = Elt | Scalar | DLEQProof
+
+// So we only encode elements, and leave packet encoding to application layer
```
Comment (sublimator) [inline src/facade/types.ts:30]: Key point

**Review by armfazh — APPROVED** [review]
lgtm, could you please squash the commits

**Comment by sublimator** [comment]
@armfazh 

Nice! Beginning to feel a lot like xmas!

**Comment by sublimator** [comment]
@thibmeu @armfazh 

Bump? :)

**Comment by thibmeu** [comment]
Merging now. It should be part of a future release.

**Comment by sublimator** [comment]
Awesome :)

**Pull request merged**